File: app/src/main/java/top/easelink/lcg/ui/main/articles/viewmodel/ArticlesViewModel.java
package top.easelink.lcg.ui.main.articles.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import java.util.List;

import top.easelink.framework.base.BaseViewModel;
import top.easelink.framework.utils.rx.SchedulerProvider;
import top.easelink.lcg.ui.main.articles.view.ArticlesNavigator;
import top.easelink.lcg.ui.main.source.model.Article;
import top.easelink.lcg.ui.main.source.remote.ArticlesRemoteDataSource;

public class ArticlesViewModel extends BaseViewModel<ArticlesNavigator> implements ArticlesAdapter.ArticlesAdapterListener  {

    private int mCurrentPage = 0;
    private String mUrl;

    private final MutableLiveData<List<Article>> articles = new MutableLiveData<>();
    private final ArticlesRemoteDataSource articlesRemoteDataSource = ArticlesRemoteDataSource.getInstance();

    public ArticlesViewModel(SchedulerProvider schedulerProvider) {
        super(schedulerProvider);
    }

    public void initUrl(String url) {
        this.mUrl = url;
        fetchArticles(FETCH_INIT);
    }

    @Override
    public void fetchArticles(int type) {
        int pageNum;
        switch (type) {
            case FETCH_MORE:
                pageNum = mCurrentPage+1;
                break;
            case FETCH_INIT:
            default:
                pageNum = 1;
                break;
        }
        setIsLoading(true);
        getCompositeDisposable().add(articlesRemoteDataSource.getHomePageArticles(mUrl, pageNum)
                .subscribeOn(getSchedulerProvider().io())
                .observeOn(getSchedulerProvider().ui())
                .subscribe(articleList -> {
                    if (articleList != null && articleList.size()!= 0) {
                        List<Article> list = articles.getValue();
                        if (type == FETCH_MORE && list != null && list.size() != 0) {
                            list.addAll(articleList);
                            articles.setValue(list);
                        } else {
                            articles.setValue(articleList);
                        }
                        // current page fetch successfully, record current page
                        mCurrentPage = pageNum;
                    }
                    setIsLoading(false);
                }, throwable -> {
                    setIsLoading(false);
                    getNavigator().handleError(throwable);
                }));
    }

    public LiveData<List<Article>> getArticles() {
        return articles;
    }

    public void scrollToTop() {
        getNavigator().scrollToTop();
    }
}

File: app/src/main/java/top/easelink/lcg/ui/main/articles/viewmodel/ArticlesViewModel.kt
package top.easelink.lcg.ui.main.articles.viewmodel

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import top.easelink.lcg.ui.main.source.model.Article
import top.easelink.lcg.ui.main.source.remote.ArticlesRemoteDataSource

class ArticlesViewModel : ViewModel(), ArticleFetcher {
    private var mCurrentPage = 0
    private var mUrl: String? = null
    val isLoading = MutableLiveData<Boolean>()
    val articles = MutableLiveData<List<Article>>()

    fun initUrl(url: String?) {
        mUrl = url
        fetchArticles(ArticleFetcher.FetchType.FETCH_INIT){}
    }

    override fun fetchArticles(fetchType: ArticleFetcher.FetchType, callback: (Boolean) -> Unit) {
        val pageNum: Int = when (fetchType) {
            ArticleFetcher.FetchType.FETCH_MORE -> mCurrentPage + 1
            ArticleFetcher.FetchType.FETCH_INIT -> 1
        }
        isLoading.value = true
        GlobalScope.launch(Dispatchers.IO){
            ArticlesRemoteDataSource.getHomePageArticles(mUrl!!, pageNum).let {
                if (it.isNotEmpty()) {
                    val list = articles.value?.toMutableList()
                    if (fetchType == ArticleFetcher.FetchType.FETCH_MORE && list != null && list.size != 0) {
                        list.addAll(it)
                        articles.postValue(list)
                    } else {
                        articles.postValue(it)
                    }
                    // current page fetch successfully, record current page
                    mCurrentPage = pageNum
                    callback.invoke(true)
                } else {
                    callback.invoke(false)
                }
            }
            isLoading.postValue(false)
        }
    }
}